server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -189,6 +192,11 @@ public IndexRequest(@Nullable ShardId shardId, StreamInput in) throws IOExceptio
                     : new ArrayList<>(possiblyImmutableExecutedPipelines);
             }
         }
+        if (in.getTransportVersion().onOrAfter(TransportVersions.REQUIRE_DATA_STREAM_ADDED)) {
+            requireDataStream = in.readBoolean();
```

Comment:
This needs to be `in.readOptionalBoolean()`, since you use `writeOptionalBoolean()` in the serialization below.